isofit/radiative_transfer/radiative_transfer.py

```diff
@@ -167,80 +166,74 @@ def coszen(self):
-                + L_up
-            )
+        # atmospheric path radiance
+        L_atm = self.get_L_atm(x_RT, geom)
 
-        elif self.topography_model:
-            I = self.solar_irr / np.pi
-            t_dir_down = r["transm_down_dir"]
-            t_dif_down = r["transm_down_dif"]
-            if geom.cos_i is None:
-                cos_i = self.coszen
-            else:
-                cos_i = geom.cos_i
-            t_total_up = r["transm_up_dif"] + r["transm_up_dir"]
-            t_total_down = t_dir_down + t_dif_down
-            s_alb = r["sphalb"]
-            # topographic flux (topoflux) effect corrected
-            ret = (
-                L_atm
-                + (
-                    I * cos_i / (1.0 - s_alb * rfl) * t_dir_down
-                    + I * self.coszen / (1.0 - s_alb * rfl) * t_dif_down
-                )
-                * rfl
-                * t_total_up
-            )
+        # atmospheric spherical albedo
+        s_alb = r["sphalb"]
 
-        elif self.glint_model:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
+        # direct and diffuse downward fluxes on the sun-to-surface path
+        # note: currently, E_down_dir comes scaled by the TOA solar zenith angle,
+        # thus, unscaling and rescaling by local solar zenith angle required
+        # to account for surface slope and aspect
+        E_down_dir, E_down_dif = self.get_E_down(x_RT, geom)
+        E_down_dir = E_down_dir / self.coszen * cos_i
 
-            E_dd = (self.solar_irr * self.coszen) / np.pi * r["t_down_dir"]
-            E_ds = (self.solar_irr * self.coszen) / np.pi * r["t_down_dif"]
-            E_d = E_dd + E_ds
-            L_sky = x_surface[-2] * E_dd + x_surface[-1] * E_ds
+        # upward direct, diffuse, and thermal transmittance
+        t_dir_up = r["transm_up_dir"]
+        t_dif_up = r["transm_up_dif"]
+        L_up = Ls * (r["transm_up_dir"] + r["transm_up_dif"])
 
+        # including glint for water surfaces
+        if self.glint_model:
             rho_ls = 0.02  # fresnel reflectance factor (approx. 0.02 for nadir view)
-            glint = rho_ls * (L_sky / E_d)
-
-            ret = (
-                L_atm
-                + L_down_transmitted
-                * (rfl + glint)
-                / (1.0 - r["sphalb"] * (rfl + glint))
-                + L_up
+            sun_glint = rho_ls * (
+                x_surface[-2] * E_down_dir / (E_down_dir + E_down_dif)
+            )
+            sky_glint = rho_ls * (
+                x_surface[-1] * E_down_dif / (E_down_dir + E_down_dif)
             )
-
         else:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
-
-            ret = L_atm + L_down_transmitted * rfl / (1.0 - r["sphalb"] * rfl) + L_up
+            sun_glint = np.zeros(rfl.shape)
+            sky_glint = np.zeros(rfl.shape)
+
+        # adjacency effects
+        bg = geom.bg_rfl if geom.bg_rfl is not None else rfl
```

Comment:
@pgbrodrick @davidraythompson In cases where we don't have `geom.bg_rfl` populated, should `bg` also contain the glint terms or should it just be `rfl`?

Reply:
Good question.  Just thinking out loud...

- glint is likely (though not always) spatially autocorrelated
- we can't know whether or not it is
- if we do include it, I think it would come into lines 219/220 now, so as to continue to separate sun and sky glint into appropriate sections (sun w/ direct, sky with diffuse....I think?)
- technically, if the bg is different, we should be solving for two more terms (background sun & sky glint), though presumably with increasing degeneracy.  At the limit though, we could enable the glint model to have 4 terms, which would be the most general way to handle it.

All that said, my default answer is probably to keep it glint-free, acknowledging that this creates a discrepancy when it's not true.  What does @davidraythompson think?

Reply:
As discussed in our meeting, I think that it may be simplest to have bg_rfl be a single vector with n values for n spectrometer channels - the effective background reflectance quantity needed for our mulitple scattering calculations.  We'll have to be careful about how we calculate this quantity, but we could start with the lambertian assumption (i.e. that the average directional reflectance we measure of the enclosing superpixel is the same everywhere).  I hypothesize this would fail in the edge case of a large placid lake acting as a specular reflector, but should get better for more typical wavy surfaces; the distribution of surface facets becomes more uniform over larger areas, right?

Reply:
Agreed, that's what I would expect as well. Another option in the frame of the lambertian assumption might be to take the prior mean from the surface model as background reflectance. In case we run ISOFIT per pixel, we wouldn't have access to the average directional reflectance from the superpixel.